src/eradiate_plugins/bsdfs/ocean_legacy.cpp

```diff
@@ -0,0 +1,1381 @@
+
+    auto [s_the, c_the]     = dr::sincos(theta);
+    auto [s_phi_w, c_phi_w] = dr::sincos(phi - wind_direction);
+
+    using FloatP        = dr::Packet<dr::scalar_t<Float>>;
+    size_t packet_count = dr::width(theta) / FloatP::Size;
+    Assert(dr::width(theta) % FloatP::Size == 0);
+
+    // cycle through each packet.
+    for (size_t i = 0; i < packet_count; ++i) {
+        FloatP s_the_p, c_the_p, s_phi_w_p, c_phi_w_p;
+        s_the_p   = dr::load<FloatP>(s_the.data() + i * FloatP::Size);
+        c_the_p   = dr::load<FloatP>(c_the.data() + i * FloatP::Size);
+        s_phi_w_p = dr::load<FloatP>(s_phi_w.data() + i * FloatP::Size);
+        c_phi_w_p = dr::load<FloatP>(c_phi_w.data() + i * FloatP::Size);
+
+        FloatP result_p = 0.f;
+        FloatP td = 0.f, summ = 0.f;
+
+        // compute the quadrature for each packet.
+        for (size_t j = 0; j < dr::width(nodes_x); ++j) {
+
+            ScalarFloat s_zenith  = s_zeniths[j];
+            ScalarFloat c_zenith  = c_zeniths[j];
+            ScalarFloat s_azimuth = s_azimuths[j];
+            ScalarFloat c_azimuth = c_azimuths[j];
+
+            ScalarFloat weight_x = weights_x[j];
+            ScalarFloat weight_y = weights_y[j];
+
+            ScalarFloat geometry       = c_zenith * s_zenith;
+            ScalarFloat geometryWeight = geometry * weight_y * weight_x;
+
+            FloatP glint = utils.eval_sun_glint(
+                n_real, n_imag, s_the_p, c_the_p, s_zenith, c_zenith, s_azimuth,
+                c_azimuth, s_phi_w_p, c_phi_w_p, upwelling);
+            td += glint * geometryWeight;
+            summ += geometryWeight;
+        }
+
+        dr::masked(td, td >= summ) = summ;
+        result_p = 1.0f - (td / summ);
+
+        dr::store(result.data() + i * FloatP::Size, result_p);
+    }
+
+    return result;
+}
+
+
+template <typename Float, typename Spectrum>
+class OceanBSDF final : public BSDF<Float, Spectrum> {
+public:
+    MI_IMPORT_BASE(BSDF, m_flags, m_components)
+    MI_IMPORT_TYPES(Texture)
+
+    /**
+     * @brief Construct a new OceanBSDF object.
+     *
+     * @param props A set of properties to initialize the oceanic BSDF.
+     */
+    OceanBSDF(const Properties &props) : Base(props) {
+        // Retrieve parameters
+        m_component      = props.get<ScalarInt32>("component", 0);
+        m_wavelength     = props.get<ScalarFloat>("wavelength");
+        m_wind_speed     = props.get<ScalarFloat>("wind_speed", 0.1f);
+        m_wind_direction = props.get<ScalarFloat>("wind_direction", 0.);
+        m_chlorinity     = props.get<ScalarFloat>("chlorinity", 19.f);
+        m_pigmentation   = props.get<ScalarFloat>("pigmentation", 0.3f);
+        m_shininess      = props.get<ScalarFloat>("shininess", 50.f);
+        m_accel          = props.get<bool>("accel", true);
+
+        update();
+
+        // Initialize the ocean utilities
+        // m_ocean_utils = new OceanUtilities<Float, Spectrum>();
```

Comment:
I suppose you rely on implicit initialization for this member?

Reply:
Yes, all of OceanUtilities' member variables have sensible defaults and a call to update() is done in OceanBSDF's constructor.